File: src/lib/models/messages.ts
// Message types for Python to Electron IPC
import { PythonChannelFunction, PythonChannel } from "./channels";
import { Result } from "./database";

// ---------- Update ----------

export interface ProgressMessage {
  step: "init";
  status: "start" | "complete" | "error";
  timestamp: number;
  updateKind: "progress";
}

export interface ExceptionMessage {
  exception: string;
  timestamp: number;
  updateKind: "exception";
}

export interface AudioLevelMessage {
  audio_level: number;
  updateKind: "audio_level";
}

export interface ErrorMessage {
  error: string;
  updateKind: "error";
}

export type ControllerStatusType =
  | "idle"
  | "recording"
  | "compressing"
  | "loading_voice_model"
  | "transcribing"
  | "loading_language_model"
  | "generating_ai_result"
  | "saving"
  | "result";

export interface StatusMessage {
  status: ControllerStatusType;
  updateKind: "status";
}

export interface ResultMessage {
  result: Result;
  updateKind: "result";
}

export interface TranscriptionMessage {
  transcription: string;
  updateKind: "transcription";
}

export type MessageType =
  | ProgressMessage
  | ExceptionMessage
  | AudioLevelMessage
  | ErrorMessage
  | StatusMessage
  | ResultMessage
  | TranscriptionMessage;

/**
 * A message from the Python process to the Electron process that was not explicitly requested.
 */
export type Update = MessageType & {
  kind: "update";
};

// ---------- Response ----------

interface BaseResponse<C extends PythonChannel> {
  data: Awaited<ReturnType<PythonChannelFunction<C>>>;
  channel: C;
  id: string;
  kind: "response";
}

/**
 * A response from the Python process to the Electron process that was explicitly requested.
 */
export type Response = {
  [C in PythonChannel]: BaseResponse<C>;
}[PythonChannel];

// ---------- Message ----------

/**
 * A message from the Python process to the Electron process.
 */
export type Message = Response | Update;

File: src-py/models/messages.py
import time
from typing import (
    Any,
    Annotated,
    Literal,
    Union,
)
from pydantic import BaseModel, Field
from models.commands import Action
from models.channels import CHANNELS
from models.db import (
    Device,
    LanguageModel,
    Mode,
    Result,
    TextReplacement,
    VoiceModel,
)

# ---------- Update Message ----------

# Utility types
StatusType = Literal["start", "complete", "error"]
StepType = Union[
    Literal["init", "recording", "compression", "transcription"],
    Action,
]


class BaseUpdateMessage(BaseModel):
    kind: Literal["update"] = "update"


class ProgressMessage(BaseUpdateMessage):
    step: StepType
    status: StatusType
    timestamp: float = Field(default_factory=time.time)
    updateKind: Literal["progress"] = "progress"


class ExceptionMessage(BaseUpdateMessage):
    exception: str
    timestamp: float = Field(default_factory=time.time)
    updateKind: Literal["exception"] = "exception"


class AudioLevelMessage(BaseUpdateMessage):
    audio_level: float
    updateKind: Literal["audio_level"] = "audio_level"


class ErrorMessage(BaseModel):
    error: str
    updateKind: Literal["error"] = "error"


ControllerStatusType = Literal[
    "idle",
    "recording",
    "compressing",
    "loading_voice_model",
    "transcribing",
    "loading_language_model",
    "generating_ai_result",
    "saving",
    "result",
]


class StatusMessage(BaseUpdateMessage):
    status: ControllerStatusType
    updateKind: Literal["status"] = "status"


class ResultMessage(BaseUpdateMessage):
    result: Result
    updateKind: Literal["result"] = "result"


class TranscriptionMessage(BaseUpdateMessage):
    transcription: str
    updateKind: Literal["transcription"] = "transcription"


# A message from the Python process to the Electron process that was not explicitly requested.
Update = Annotated[
    Union[
        ProgressMessage,
        ExceptionMessage,
        AudioLevelMessage,
        ErrorMessage,
        StatusMessage,
        ResultMessage,
        TranscriptionMessage,
    ],
    Field(discriminator="updateKind"),
]


# ---------- Response Message ----------


class BaseResponse(BaseModel):
    """
    Base class for a response from the Python process to the Electron process that was explicitly requested.
    """

    data: Any  # data type depends on the channel, refer to awaited return types of PythonChannelMap in channels.ts. Subtypes implement tighter types.
    channel: CHANNELS
    id: str
    kind: Literal["response"] = "response"


# TODO: investigate the need for the channel field


class ModesResponse(BaseResponse):
    data: list[Mode]
    channel: Literal[CHANNELS.FETCH_ALL_MODES] = CHANNELS.FETCH_ALL_MODES


class ModeResponse(BaseResponse):
    data: Mode
    channel: Literal[CHANNELS.CREATE_MODE] = CHANNELS.CREATE_MODE


class ResultsResponse(BaseResponse):
    data: list[Result]
    channel: Literal[CHANNELS.FETCH_ALL_RESULTS] = CHANNELS.FETCH_ALL_RESULTS


class TextReplacementsResponse(BaseResponse):
    data: list[TextReplacement]
    channel: Literal[CHANNELS.FETCH_ALL_TEXT_REPLACEMENTS] = (
        CHANNELS.FETCH_ALL_TEXT_REPLACEMENTS
    )


class DevicesResponse(BaseResponse):
    data: list[Device]
    channel: Literal[CHANNELS.FETCH_ALL_DEVICES] = CHANNELS.FETCH_ALL_DEVICES


class VoiceModelsResponse(BaseResponse):
    data: list[VoiceModel]
    channel: Literal[CHANNELS.FETCH_ALL_VOICE_MODELS] = CHANNELS.FETCH_ALL_VOICE_MODELS


class LanguageModelsResponse(BaseResponse):
    data: list[LanguageModel]
    channel: Literal[CHANNELS.FETCH_ALL_LANGUAGE_MODELS] = (
        CHANNELS.FETCH_ALL_LANGUAGE_MODELS
    )


# ------- Message -------

# A message from the Python process to the Electron process.
Message = Annotated[
    Union[BaseResponse, Update],
    Field(discriminator="kind"),
]
